src/webgpu/shader/validation/uniformity/uniformity.spec.ts

```diff
@@ -1607,6 +1607,239 @@ const kShortCircuitExpressionCases = {
+    var y = uniform_values[1];
+    foo(&x, &y);
+    let a = x + y;`,
+    cond: `a > 0`,
+    uniform: true,
+  },
+  pointer_codependent2: {
+    function: `fn foo(p : ptr<function, u32>, q : ptr<function, u32>) {
+      if *p > 0 {
+        *p = *q;
+      } else {
+        *q++;
+      }
+    }`,
+    call: `var x = uniform_values[0];
+    var y = nonuniform_values[1];
+    foo(&x, &y);
+    let a = x + y;`,
+    cond: `a > 0`,
+    uniform: false,
+  },
+  pointer_codependent3: {
+    function: `fn foo(p : ptr<function, u32>, q : ptr<function, u32>) {
+      if *p > 0 {
+        *p = *q;
+      } else {
+        *q++;
+      }
+    }`,
+    call: `var x = nonuniform_values[0];
+    var y = uniform_values[1];
+    foo(&x, &y);
+    let a = x + y;`,
+    cond: `a > 0`,
+    uniform: false,
+  },
+  pointer_codependent4: {
+    function: `fn foo(p : ptr<function, u32>, q : ptr<function, u32>) {
+      if *p > 0 {
+        *p = *q;
+      } else {
+        *q++;
+      }
+    }`,
+    call: `var x = nonuniform_values[0];
+    var y = nonuniform_values[1];
+    foo(&x, &y);
+    let a = x + y;`,
+    cond: `a > 0`,
+    uniform: false,
+  },
+  uniform_param_uniform_assignment: {
+    function: `fn foo(p : ptr<function, array<u32, 2>>, idx : u32) {
+      (*p)[idx] = uniform_values[0];
+    }`,
+    call: `var x = array(uniform_values[0], uniform_values[1]);
+    foo(&x, uniform_values[3]);`,
+    cond: `x[0] > 0`,
+    uniform: true,
+  },
+  uniform_param_nonuniform_assignment: {
+    function: `fn foo(p : ptr<function, array<u32, 2>>, idx : u32) {
+      (*p)[idx] = nonuniform_values[0];
+    }`,
+    call: `var x = array(uniform_values[0], uniform_values[1]);
+    foo(&x, uniform_values[3]);`,
+    cond: `x[0] > 0`,
+    uniform: false,
+  },
+  nonuniform_param_uniform_assignment: {
+    function: `fn foo(p : ptr<function, array<u32, 2>>, idx : u32) {
+      (*p)[idx] = uniform_values[0];
+    }`,
+    call: `var x = array(uniform_values[0], uniform_values[1]);
+    foo(&x, u32(clamp(pos.x, 0, 1)));`,
+    cond: `x[0] > 0`,
+    uniform: false,
+  },
+  nonuniform_param_nonuniform_assignment: {
+    function: `fn foo(p : ptr<function, array<u32, 2>>, idx : u32) {
+      (*p)[idx] = nonuniform_values[0];
+    }`,
+    call: `var x = array(uniform_values[0], uniform_values[1]);
+    foo(&x, u32(clamp(pos.x, 0, 1)));`,
+    cond: `x[0] > 0`,
+    uniform: false,
+  },
+  required_uniform_success: {
+    function: `fn foo(p : ptr<function, u32>) {
+      if *p > 0 {
+        let tmp = textureSample(t,s,vec2f(0,0));
+      }
+    }`,
+    call: `var x = uniform_values[0];
+    foo(&x);`,
+    cond: `uniform_cond`,
+    uniform: true,
+  },
+  required_uniform_failure: {
+    function: `fn foo(p : ptr<function, u32>) {
+      if *p > 0 {
+        let tmp = textureSample(t,s,vec2f(0,0));
+      }
+    }`,
+    call: `var x = nonuniform_values[0];
+    foo(&x);`,
+    cond: `uniform_cond`,
+    uniform: false,
+  },
+};
```

Comment:
One graph edge that I don't see tested here is a conditional call to a function that assigns through a pointer. e.g.
```
fn foo(p : ptr<function, u32>) {
  *p = uniform_values[0];
}
...
var v : u32;
if (non_uniform) {
  foo(&v);
}
if (v > 0) {
  let tmp = textureSample(...);
}
```
This tests that there's an edge from the value that the pointer has after the function call to the incoming control flow of the function call. We missed this in the initial implementation in Tint and it was found later.

Reply:
Added a few cases along those lines.